### PR TITLE
fix #129 install process listeners safely

### LINKED ISSUE #129: tmp installs process listeners multiple times
## Operating System
- all OS

## NodeJS Version
- all versions 

## Tmp Version
all versions

## Expected Behavior
tmp must not try to install process exit and uncaughtException listeners multiple times.

## Experienced Behavior
tmp installs listeners multiple times, causing a warning "possible event listener/memory leak".
This occurs when other projects use facebook/jest which removes and thus reloads existing modules during each test run. And it might also occur when using node's vm (sandboxed modules) API.

## Rationale

Personally, I believe that facebook/jest is doing it wrong, but for the time being we should provide a work around for that problem.

As for node's sandbox VM module, we need to have additional tests that make sure that in such a scenario the listeners will not be added multiple times. Implementing such a test case properly will
require preventing tmp to install its listeners safely. Since this cannot be done, we will simply make sure that in a multiple sandbox environment the aforementioned warning will not be triggered.   

In addition, we should notify the owners of dependent other projects of the update.
And, by analyzing the package dependencies of https://github.com/chrisprice/react-app-webdriver, the most prominent would be

- makeerror
- selenium-webdriver

## See Also

#125

### PR DESCRIPTION
**Pull request by silkentrance**
See #129 

WIP
TODO: additional tests

~~~In order for the tests to run without error this requires #135 which has been merged.~~~

### CODE REVIEW

File: lib/tmp.js [hunk 1]
```diff
@@ -9,12 +9,18 @@
...
 const fs = require('fs');
 const path = require('path');
 const crypto = require('crypto');
-const osTmpDir = require('os-tmpdir');
 const _c = process.binding('constants');
 
+
+// we use this for determining which SIGINT handler we need to install
+// TODO:is M$ so dim, win32 for win64 and future win128 platforms?
+const isWindows = process.platform == 'win32';
```
Comment (raszi) [inline lib/tmp.js]: This is unused in this PR.

File: lib/tmp.js [hunk 2]
```diff
@@ -96,15 +102,14 @@ function _isUndefined(obj) {
...
 function _parseArguments(options, callback) {
-  if (typeof options == 'function') {
-    return [callback || {}, options];
-  }
+  var result = [options || {}, callback]; 
 
-  if (_isUndefined(options)) {
-    return [{}, callback];
+  if (typeof options == 'function') {
+    result[0] = {};
```
Comment (raszi) [inline lib/tmp.js]: I prefer the early returns and the immutable values instead of this.

File: lib/tmp.js [hunk 3]
```diff
@@ -9,12 +9,18 @@
 /*
  * Module dependencies.
  */
+const tmpdir = require('os-tmpdir');
```
Comment (raszi) [inline lib/tmp.js]: Why was it necessary to rename the const?

File: lib/tmp.js [hunk 4]
```diff
@@ -517,19 +525,54 @@ const version = process.versions.node.split('.').map(function (value) {
...
+function _is_legacy_listener(name) {
+  return name == '_exit' || name == '_uncaughtExceptionThrown';
+}
+
+function _safely_install_listener(evt, listener) {
+  var listeners = process.listeners(evt);
+
+  for (var i = 0, length = listeners.length; i < length; i++) {
+    var lstnr = listeners[i];
+    if ((lstnr.name == listener.name) || _is_legacy_listener(listener.name)) {
```
Comment (raszi) [inline lib/tmp.js]: No need for the extra parens around the equality check.

File: lib/tmp.js [hunk 5]
```diff
@@ -154,6 +160,7 @@ function tmpName(options, callback) {
   if (opts.template && !opts.template.match(TEMPLATE_PATTERN))
     return cb(new Error('Invalid template provided'));
 
+  // FIXME:get rid of duplicate code
```
Comment (raszi) [inline lib/tmp.js]: Please put a space after the FIXME keyword, like this:

```javascript
// FIXME get rid of duplicate code
```

File: lib/tmp.js (hunk 3 above)
Comment (silkentrance) [inline lib/tmp.js]: fixed

File: lib/tmp.js (hunk 1 above)
Comment (silkentrance) [inline lib/tmp.js]: fixed

File: lib/tmp.js (hunk 2 above)
Comment (silkentrance) [inline lib/tmp.js]: fixed

File: lib/tmp.js (hunk 5 above)
Comment (silkentrance) [inline lib/tmp.js]: fixed

File: lib/tmp.js (hunk 4 above)
Comment (silkentrance) [inline lib/tmp.js]: fixed

**Comment by silkentrance** [comment]
NOTE that this is still work in progress

File: lib/tmp.js [hunk 6]
```diff
@@ -97,14 +97,14 @@ function _isUndefined(obj) {
...
-    return [callback || {}, options];
+    return [{}, options];
   }
 
   if (_isUndefined(options)) {
     return [{}, callback];
   }
 
-  return [options, callback];
+  return [options, callback];
```
Comment (raszi) [inline lib/tmp.js]: There is an unnecessary whitespace at the end of the line.

Comment (silkentrance) [inline lib/tmp.js]: fixed, but not on this line :grin:, there was an extra WS in line 271 after the fs.closeSync(); call.

**Comment by silkentrance** [comment]
IMPORTANT tests are still missing!

**Comment by silkentrance** [comment]
There seems to be an issue with the new tests. I will have to look into that.

**Comment by silkentrance** [comment]
@raszi while the existing test case test/issue129-test and test/outband/issue129.js(on) does not completely encompass the behaviour of facebook/jest, which utilises sandboxes, I think that we can be safe in the assumption that tmp will register its handlers safely and not try to register handlers multiple times.

That being said, I dropped the uncaughtException handler completely. And while earlier versions of tmp will still install that handler, I think that we can be safe to assume that it does not matter.

However, we will still have to notify interested third parties to update their dependencies accordingly.

**Pull request merged**